### ISSUE TITLE
should inverse_transform always return 4D output?

### ISSUE THREAD

**Issue by jdkent** [nilearn/nilearn#2726]
The controversy as I see it is balancing the philosophical choices of the maintainers versus how the code has actually been behaving over the last near decade (and how many people depend on that behavior).

> @tsalo, great point you brought up, for `inverse_transform`, if a 1D input is passed in then a 3D output will be produced, but looking at https://github.com/nilearn/nilearn/issues/460#issuecomment-100767188, I see that consistency of output was deemed more important, if that decision still holds, `inverse_transform` should always return 4D. Although, the functionality for unmask was [implemented before that conversation](https://github.com/nilearn/nilearn/pull/42).
\
I do not have a strong opinion on this (but someone should!), for maintainability I believe the 2D/4D output should be enforced, but on the other hand unmask has supported 1D inputs for years suggesting people may depend on that behavior of unmask and the transformers that call unmask.
\
I exploited the idiosyncrasy of `inverse_transform` for [NiMARE](https://github.com/neurostuff/NiMARE/blob/713a231527b3d4615a86161bcf174e0a52c96752/nimare/transforms.py#L123), but that behavior may be a bug and not a feature.

_Originally posted by @jdkent in https://github.com/nilearn/nilearn/pull/2725#issuecomment-794386153_

**Comment by jdkent** [comment]
also relevent:


> Enforcing 4D output implies a bug was introduced ~8 years ago, when `unmask` allowed [1D inputs and returned 3D outputs](https://github.com/nilearn/nilearn/pull/42/files#diff-ec47025e5fd3f9fb5ad16ef77a0e6aef29507c171ff3720c5c0cd6f9a4e9ba42R398-R401), and the `unmask` function was called in [base_masker.py under the `inverse_transform` method](https://github.com/nilearn/nilearn/blob/378d9ddd96c3e02ceeee7494863a6616d4e227f5/nisl/io/base_masker.py#L143).
> 
> In proper context, my usage of `inverse_transform` on 1D input and getting 3D output in NiMARE is a long-standing bug (not a feature), that should be fixed.
>

**Comment by jdkent** [comment]
### inverse_transforms that do ensure 4D output: (uses `signal_extraction` in `inverse_transform`)

- Parcellations
- NiftiSpheresMasker
- NiftiLabelsMasker


### inverse_transforms that do not ensure 4D output: (uses `unmask` in `inverse_transform` directly or indirectly)

- NiftiMapsMasker
- MultiNiftiMasker
- NiftiMasker
- RegionExtractor

### always 2D

- ReNA

ReNA should probably remain unchanged, but I (naively?) expect the other maskers/extractors to have consistent behavior when passed 1D input into `inverse_transform`.

**Comment by jdkent** [comment]
@NicolasGensollen @bthirion WDYT?

For enforcement of 4D outputs:
- @bthirion:  https://github.com/nilearn/nilearn/pull/2725#issuecomment-794399854 
- @GaelVaroquaux (2015): https://github.com/nilearn/nilearn/issues/460#issuecomment-100767188

Against enforcement of 4D outputs:
- ???


(should others be tagged in this conversation?)

**Comment by NicolasGensollen** [comment]
Thanks for opening this and listing these inconsistencies @jdkent :+1: 
I would be in favour of consistency across maskers, meaning they all accept the same input format(s) (2D or 1D/2D) and output the same format as well (4D).
However, changing the return value of `inverse_transform` when passing 1D inputs from 3D to 4D is a breaking change as this is likely to break dependent codes that expect a 3D output. I think such a change would require a deprecation cycle to let people update their code. 
Tagging also @jeromedockes who might have an opinion on this.

**Comment by NicolasGensollen** [comment]
>Namely, 3D inputs to transform produce 1D outputs, and 1D inputs to inverse_transform produce 3D outputs?

I don't think this would cause backward compatibility issues indeed, but changing the output of `inverse_transform` from 3D to 4D for 1D inputs would (for the `NiftiMasker` at least since the others don't support it yet). 
So, I think it all depends if we want to enforce systematic 4D outputs for `inverse_transform`. 

I'd be in favour of enforcing 4D outputs for `inverse_transform` because `transform` always outputs 2D whatever the input dimension (3D/4D), so I'd expect `inverse_transform` to have the same consistent behaviour.

That being said, if we enforce 4D outputs for `inverse_transform`, it means that doing `inverse_transform(transform(X))` with a 3D X input will not give you back a 3D array, which might be confusing for some users. But this is already the case for the `NiftiMasker` anyway...

```python
import numpy as np
from nilearn.input_data import NiftiMasker
from nibabel import Nifti1Image

rng = np.random.RandomState(42)
mask = np.zeros((10, 10, 10))
mask[3:7, 3:7, 3:7] = 1
mask_img = Nifti1Image(mask, np.eye(4))
masker = NiftiMasker(mask_img=mask_img)

data_4D = Nifti1Image(rng.random_sample((10, 10, 10, 3)), np.eye(4))
# 4D transform gives 2D
signals = masker.fit_transform(data_4D)
assert signals.shape == (3, 64)
# 2D inverse_transform gives 4D
inv_out = masker.inverse_transform(signals)
assert inv_out.shape == data_4D.shape

data_3D = Nifti1Image(rng.random_sample((10, 10, 10)), np.eye(4))
# 3D transform gives 2D
signals = masker.fit_transform(data_3D)
assert signals.shape == (1, 64)
# 2D inverse_transform gives 4D
inv_out = masker.inverse_transform(signals)
assert inv_out.shape == (10, 10, 10, 1)
# 1D inverse_transform gives 3D
inv_out2 = masker.inverse_transform(signals.ravel())
assert inv_out2.shape == (10, 10, 10)
```

**Comment by bthirion** [comment]
> I'd be in favour of enforcing 4D outputs for `inverse_transform` because `transform` always outputs 2D whatever the input dimension (3D/4D), so I'd expect `inverse_transform` to have the same consistent behaviour.

+1 I think that this is more consistent overall. But we need a wider consensus on that question.

**Comment by jdkent** [comment]
If you are tagged, this means you've worked on a repository that uses `inverse_transform` from nilearn.
The question we are trying to answer is:

**If one passes a 1D input into the `inverse_transform` method of a masker, do you expect the output to be 3D or 4D?**

context: Instead of inverting multiple samples/flattened images (2D input), you are only inverting one sample/flattened image (1D input), should the output be a 4D nibabel image or a 3D nibabel image?
- :tada:: I expect 3D output
- 🚀 : I expect 4D output
- 👀 : seen, do not care

React to this comment with the above emoji that reflects your opinion/how you use `inverse_transform` in your code.
(please comment if your opinion does not fit in an emoji 😄)

people that may have opinions:
- @jeromedockes (already tagged previously): https://github.com/neuroquery/neuroquery
- @thomasbazeille, @emdupre: https://github.com/Parietal-INRIA/fmralign
- @calebgeniesse, @RdoubleA, @jsmentch: https://github.com/braindynamicslab/dyneusr
- @lukassnoek: https://github.com/lukassnoek/skbold
- @ejolly, @ljchang: https://github.com/cosanlab/nltools
- @effigies, @adelavega: https://github.com/poldracklab/fitlins
- @tsalo: https://github.com/neurostuff/NiMARE

(command used to find relevant people: `ghtopdep https://github.com/nilearn/nilearn --packages --rows=5 --search=inverse_transform`) from the [ghtopdep repo](https://github.com/github-tooling/ghtopdep))

**Comment by adelavega** [comment]
I answered based on what I would naiively expect, but given the conversation, I also expect uniform behavior across package, and would deem that more important (i.e. 4D is more consistent)

**Comment by jdkent** [comment]
some relevant discussions from scikit-learn:
- scikit-learn/scikit-learn#4509
- scikit-learn/scikit-learn#4511
- scikit-learn/scikit-learn#4252

Moving this conversation from just outputs to inputs and outputs, I see three options:
- transform(4D) -> 2D output; inverse_transform(2D) -> 4D output
  - PROS:
    - most compatible with scikit-learn
    - do not need to deal with ambiguity between `(n_samples,)` or `(n_features,)`
    - likely best option for maintainable code
  - CONS:
    - makes common non scikit-learn based usage of nilearn less user friendly
  - what the "common" non scikit-learn use case could look like:
    - `arr_1d = masker.transform([img_3d])[:, 0]` (place niftiimage in list to make it "4D" and only select the first sample to make the output 1D)
    - `img_3d = nibabel.funcs.sqeeze_image(masker.inverse_transform(arr_1D[:, np.newaxis]))` (add axis to make arr_1D a 2D array, and squeeze the output to make the output 3D)
- transform(3D or 4D) -> 2D output; inverse_transform(1D or 2D) -> 4D output
  - PROS:
    - can still be used in scikit-learn pipelines
    - covers common non scikit-learn use case of reading in a single niftiimage file to work with masked data
  - CONS:
    - potential ambiguity of  `(n_samples,)` or `(n_features,)` (although I do not see this as a problem for nilearn specifically)
    - need to maintain more complex code
  - what the "common" non scikit-learn use case could look like:
    - `arr_1d = masker.transform(img_3d)[:, 0]` (select the first sample to make the output 1D)
    - `img_3d = nibabel.funcs.sqeeze_image(masker.inverse_transform(arr_1D))` (squeeze the output to make the output 3D)
- transform(3D or 4D) -> 1D or 2D output; inverse_transform(1D or 2D) -> 3D or 4D output
  - PROS
    - covers common non scikit-learn usecase of reading in a single niftiimage file to work with masked data
    - covers common non scikit-learn usecase of writing out a 1D numpy array to a nifti image
  - CONS:
    - incompatible with scikit-learn estimators (which always expect 2D)
    - most complex code to maintain
  - what the "common" non scikit-learn use case could look like:
    - `arr_1d = masker.transform(img_3d)`
    - `img_3d = masker.inverse_transform(arr_1D))`
  
From the above, I would say the middle option looks best to compromise the maintainability of code used in nilearn
and the maintainability of code used in other packages that use the maskers for non scikit-learn purposes (from the perspective of someone that uses nilearn maskers for non scikit-learn purposes).

I could also see choosing option 1 as perfectly reasonable (as this is what scikit-learn chose), but it may alienate the non scikit-learn use cases/audience (which you may not wish to cater to in the first place).


## Rambling

Using the wisdom from those conversations, 2D/4D inputs/outputs is all nilearn should accept.
using 1D input led to ambiguity whether the array represented `(n_samples,)` for one feature or `(n_features,)` for one sample.

What's different in nilearn (relative to scikit-learn) is the types of inputs are more constrained (AFAICT). scikit-learn has to be able to handle inputs representing a myriad of things (text from speech, audio signals, images), where there is really no right answer on whether a 1D input represents `(n_samples,)` or `(n_features,)`.

nilearn, on the other hand, handles brain images with many features (voxels, vertexes, etc.), so it could be reasoned a sensible default for 1D input is `(n_features,)`, which means we can reasonably interpret 3D output from `inverse_transform`.

Another factor is whether nilearn wants to stay within the framework of scikit-learn/machine learning or accept a seemingly common use case of transforming single images into 1D output in order to apply operations over that array in an efficient manner before inverting the transformation back into a single image.
A critical downside of accepting the aforementioned seemingly common use case is lack of interoperability of nilearn and scikit-learn (i.e., passing in a 3D image and getting 1D output could not be passed into any of scikit-learn's estimators)

nilearn provides a powerful/elegant way of transforming single images which it appears a few developers have taken advantage of, and the ambiguity of `(n_samples,)` or `(n_features,)` appears to be less of an issue in nilearn (to me at least, happy to be shown otherwise).

**Comment by bthirion** [comment]
Thx for the complete discussion.
I think I prefer suggestion 2. But we need a wide agreement on this. Can people voice their opinion here ? Or should we keep this for a live discussion during the upcoming Nilearn sprint (https://nilearn.github.io/dev-days-2021) ?

**Comment by tsalo** [comment]
Here's a relevant note from the [dev call in June](https://github.com/nilearn/admin/blob/main/meeting_notes/2022/07_06_2022.rst):

>Consensus is that it would be reasonable to return 3D outputs when it is clear that there is no sample dimension (though there are probably backwards-incompatibility issues to deal with).

I suppose the question is whether we want to implement this now or have a deprecation cycle for it.

**Comment by htwangtw** [comment]
Just a side note - while doing some performance benchmark on all the maskers, I might have found some potentially undesired inverse_transform behaviour...
See notes on this commit on my personal fork:
https://github.com/htwangtw/nilearn/tree/4ab07bce50e10cdd7627cfc487720527df455972

The bench mark I have yet got time to revisit:
https://github.com/SIMEXP/mapsmasker_benchmark

**Comment by tsalo** [comment]
Oh, that's interesting. Is it worth opening a new bug report for the NiftiMapsMasker's resampling approach?

**Comment by bthirion** [comment]
Thx for raising this.

**Comment by tsalo** [comment]
I think we should probably have a deprecation cycle where a warning is raised when 3D niimgs are used in the `transform` method of those maskers that currently enforce 2D outputs, and when 1D arrays are used in the `inverse_transform` method of those maskers that currently enforce 4D outputs.

Would two minor releases (e.g., `0.10.0` and `0.11.0`) be enough? I.e., make the change in `0.12.0`?

**Comment by bthirion** [comment]
Sounds reasonable.
